File: src/main/java/alfheim/client/render/tile/RenderTileAnomaly.java
package alfheim.client.render.tile;

import static org.lwjgl.opengl.GL11.*;

import alfheim.api.block.tile.SubTileEntity;
import alfheim.common.block.tile.TileAnomaly;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.ActiveRenderInfo;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.tileentity.TileEntitySpecialRenderer;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.MathHelper;
import net.minecraft.util.Vec3;

// Render from Thaumcraft nodes by Azanor
public class RenderTileAnomaly extends TileEntitySpecialRenderer {
	
	public void renderTileEntityAt(TileEntity tile, double x, double y, double z, float partialTicks) {
		if (!(tile instanceof TileAnomaly)) return;
		TileAnomaly anomaly = (TileAnomaly) tile;
		
		SubTileEntity mainSTE = anomaly.subTiles.get(anomaly.mainSubTile);
		if (mainSTE == null) return;
		
		float pt = Minecraft.getMinecraft().timer.renderPartialTicks;
		mainSTE.bindTexture();
		
		glPushMatrix();
		glAlphaFunc(GL_GREATER, 0.003921569F);
		glEnable(GL_BLEND);
		glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
		glDisable(GL_LIGHTING);
		glDepthMask(false);
		
		glTranslated(mainSTE.x() + 0.5, mainSTE.y() + 0.5, mainSTE.z() + 0.5);
		
		int frame = (int) ((System.nanoTime() / 40000000L + x) % mainSTE.getFrames());
		
		renderFacingStrip(0, 0, 0, 0, 1, 1, mainSTE.getFrames(), mainSTE.getStrip(), frame, partialTicks, mainSTE.getColor());
		
		glDepthMask(true);
		glEnable(GL_LIGHTING);
		glDisable(GL_BLEND);
		glAlphaFunc(GL_GREATER, 0.1F);
		glPopMatrix();
		
	}
	
	public static void renderFacingStrip(double px, double py, double pz, float angle, float scale, float alpha, int frames, int strip, int frame, float partialTicks, int color) {
		if (Minecraft.getMinecraft().renderViewEntity instanceof EntityPlayer) {
			Tessellator tessellator = Tessellator.instance;
			float arX = ActiveRenderInfo.rotationX;
			float arZ = ActiveRenderInfo.rotationZ;
			float arYZ = ActiveRenderInfo.rotationYZ;
			float arXY = ActiveRenderInfo.rotationXY;
			float arXZ = ActiveRenderInfo.rotationXZ;
			EntityPlayer player = (EntityPlayer) Minecraft.getMinecraft().renderViewEntity;
			double iPX = player.prevPosX + (player.posX - player.prevPosX) * (double) partialTicks;
			double iPY = player.prevPosY + (player.posY - player.prevPosY) * (double) partialTicks;
			double iPZ = player.prevPosZ + (player.posZ - player.prevPosZ) * (double) partialTicks;
			glTranslated(-iPX, -iPY, -iPZ);
			tessellator.startDrawingQuads();
			tessellator.setBrightness(220);
			tessellator.setColorRGBA_I(color, (int) (alpha * 255.0F));
			Vec3 v1 = Vec3.createVectorHelper((double) (-arX * scale - arYZ * scale), (double) (-arXZ * scale), (double) (-arZ * scale - arXY * scale));
			Vec3 v2 = Vec3.createVectorHelper((double) (-arX * scale + arYZ * scale), (double) (arXZ * scale), (double) (-arZ * scale + arXY * scale));
			Vec3 v3 = Vec3.createVectorHelper((double) (arX * scale + arYZ * scale), (double) (arXZ * scale), (double) (arZ * scale + arXY * scale));
			Vec3 v4 = Vec3.createVectorHelper((double) (arX * scale - arYZ * scale), (double) (-arXZ * scale), (double) (arZ * scale - arXY * scale));
			if (angle != 0.0F) {
				Vec3 f2 = Vec3.createVectorHelper(iPX, iPY, iPZ);
				Vec3 f3 = Vec3.createVectorHelper(px, py, pz);
				Vec3 f4 = f2.subtract(f3).normalize();
				QuadHelper.setAxis(f4, (double) angle).rotate(v1);
				QuadHelper.setAxis(f4, (double) angle).rotate(v2);
				QuadHelper.setAxis(f4, (double) angle).rotate(v3);
				QuadHelper.setAxis(f4, (double) angle).rotate(v4);
			}
			
			float f21 = (float) frame / (float) frames;
			float f31 = (float) (frame + 1) / (float) frames;
			float f41 = (float) strip / (float) frames;
			float f5 = ((float) strip + 1.0F) / (float) frames;
			tessellator.setNormal(0.0F, 0.0F, -1.0F);
			tessellator.addVertexWithUV(px + v1.xCoord, py + v1.yCoord, pz + v1.zCoord, (double) f31, (double) f5);
			tessellator.addVertexWithUV(px + v2.xCoord, py + v2.yCoord, pz + v2.zCoord, (double) f31, (double) f41);
			tessellator.addVertexWithUV(px + v3.xCoord, py + v3.yCoord, pz + v3.zCoord, (double) f21, (double) f41);
			tessellator.addVertexWithUV(px + v4.xCoord, py + v4.yCoord, pz + v4.zCoord, (double) f21, (double) f5);
			tessellator.draw();
		}
	}
	
	private static class QuadHelper {
		
		public final double x;
		public final double y;
		public final double z;
		public final double angle;
		
		public QuadHelper(double ang, double xx, double yy, double zz) {
			this.x = xx;
			this.y = yy;
			this.z = zz;
			this.angle = ang;
		}
		
		public static QuadHelper setAxis(Vec3 vec, double angle) {
			angle *= 0.5D;
			double d4 = (double) MathHelper.sin((float) angle);
			return new QuadHelper((double) MathHelper.cos((float) angle), vec.xCoord * d4, vec.yCoord * d4, vec.zCoord * d4);
		}
		
		public void rotate(Vec3 vec) {
			double d = -this.x * vec.xCoord - this.y * vec.yCoord - this.z * vec.zCoord;
			double d1 = this.angle * vec.xCoord + this.y * vec.zCoord - this.z * vec.yCoord;
			double d2 = this.angle * vec.yCoord - this.x * vec.zCoord + this.z * vec.xCoord;
			double d3 = this.angle * vec.zCoord + this.x * vec.yCoord - this.y * vec.xCoord;
			vec.xCoord = d1 * this.angle - d * this.x - d2 * this.z + d3 * this.y;
			vec.yCoord = d2 * this.angle - d * this.y + d1 * this.z - d3 * this.x;
			vec.zCoord = d3 * this.angle - d * this.z - d1 * this.y + d2 * this.x;
		}
	}
}

File: src/main/java/alfheim/client/render/tile/RenderTileAnomaly.kt
package alfheim.client.render.tile

import alexsocol.asjlib.*
import alfheim.common.block.tile.TileAnomaly
import net.minecraft.client.renderer.*
import net.minecraft.client.renderer.tileentity.TileEntitySpecialRenderer
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.Vec3
import org.lwjgl.opengl.GL11.*
import kotlin.math.*

// Render from Thaumcraft nodes by Azanor
object RenderTileAnomaly: TileEntitySpecialRenderer() {
	
	override fun renderTileEntityAt(tile: TileEntity, x: Double, y: Double, z: Double, partialTicks: Float) {
		if (tile !is TileAnomaly) return
		
		val mainSTE = tile.subTiles[tile.mainSubTile] ?: return
		
		mainSTE.bindTexture()
		
		glPushMatrix()
		glAlphaFunc(GL_GREATER, 0.003921569f)
		glEnable(GL_BLEND)
		glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)
		glDisable(GL_LIGHTING)
		glDepthMask(false)
		
		glTranslated(mainSTE.x() + 0.5, mainSTE.y() + 0.5, mainSTE.z() + 0.5)
		
		val frame = System.nanoTime().div(40000000L).plus(x).toLong().rem(mainSTE.frames).I
		
		renderFacingStrip(0.0, 0.0, 0.0, 0f, 1f, 1f, mainSTE.frames, mainSTE.strip, frame, partialTicks, mainSTE.color)
		
		glDepthMask(true)
		glEnable(GL_LIGHTING)
		glDisable(GL_BLEND)
		glAlphaFunc(GL_GREATER, 0.1f)
		glPopMatrix()
		
	}
	
	private class QuadHelper(val angle: Double, val x: Double, val y: Double, val z: Double) {
		
		fun rotate(vec: Vec3) {
			val d = -x * vec.xCoord - y * vec.yCoord - z * vec.zCoord
			val d1 = angle * vec.xCoord + y * vec.zCoord - z * vec.yCoord
			val d2 = angle * vec.yCoord - x * vec.zCoord + z * vec.xCoord
			val d3 = angle * vec.zCoord + x * vec.yCoord - y * vec.xCoord
			vec.xCoord = d1 * angle - d * x - d2 * z + d3 * y
			vec.yCoord = d2 * angle - d * y + d1 * z - d3 * x
			vec.zCoord = d3 * angle - d * z - d1 * y + d2 * x
		}
		
		companion object {
			
			fun setAxis(vec: Vec3, ang: Double): QuadHelper {
				val angle = ang * 0.5
				val d4 = sin(angle)
				return QuadHelper(cos(angle), vec.xCoord * d4, vec.yCoord * d4, vec.zCoord * d4)
			}
		}
	}
	
	fun renderFacingStrip(px: Double, py: Double, pz: Double, angle: Float, scale: Float, alpha: Float, frames: Int, strip: Int, frame: Int, partialTicks: Float, color: Int) {
		if (mc.renderViewEntity is EntityPlayer) {
			val tessellator = Tessellator.instance
			val arX = ActiveRenderInfo.rotationX
			val arZ = ActiveRenderInfo.rotationZ
			val arYZ = ActiveRenderInfo.rotationYZ
			val arXY = ActiveRenderInfo.rotationXY
			val arXZ = ActiveRenderInfo.rotationXZ
			val player = mc.renderViewEntity as EntityPlayer
			val iPX = player.prevPosX + (player.posX - player.prevPosX) * partialTicks.D
			val iPY = player.prevPosY + (player.posY - player.prevPosY) * partialTicks.D
			val iPZ = player.prevPosZ + (player.posZ - player.prevPosZ) * partialTicks.D
			glTranslated(-iPX, -iPY, -iPZ)
			tessellator.startDrawingQuads()
			tessellator.setBrightness(220)
			tessellator.setColorRGBA_I(color, (alpha * 255f).I)
			val v1 = Vec3.createVectorHelper((-arX * scale - arYZ * scale).D, (-arXZ * scale).D, (-arZ * scale - arXY * scale).D)
			val v2 = Vec3.createVectorHelper((-arX * scale + arYZ * scale).D, (arXZ * scale).D, (-arZ * scale + arXY * scale).D)
			val v3 = Vec3.createVectorHelper((arX * scale + arYZ * scale).D, (arXZ * scale).D, (arZ * scale + arXY * scale).D)
			val v4 = Vec3.createVectorHelper((arX * scale - arYZ * scale).D, (-arXZ * scale).D, (arZ * scale - arXY * scale).D)
			if (angle != 0f) {
				val f2 = Vec3.createVectorHelper(iPX, iPY, iPZ)
				val f3 = Vec3.createVectorHelper(px, py, pz)
				val f4 = f2.subtract(f3).normalize()
				QuadHelper.setAxis(f4, angle.D).rotate(v1)
				QuadHelper.setAxis(f4, angle.D).rotate(v2)
				QuadHelper.setAxis(f4, angle.D).rotate(v3)
				QuadHelper.setAxis(f4, angle.D).rotate(v4)
			}
			
			val f21 = frame.F / frames.F
			val f31 = (frame + 1).F / frames.F
			val f41 = strip.F / frames.F
			val f5 = (strip.F + 1f) / frames.F
			tessellator.setNormal(0f, 0f, -1f)
			tessellator.addVertexWithUV(px + v1.xCoord, py + v1.yCoord, pz + v1.zCoord, f31.D, f5.D)
			tessellator.addVertexWithUV(px + v2.xCoord, py + v2.yCoord, pz + v2.zCoord, f31.D, f41.D)
			tessellator.addVertexWithUV(px + v3.xCoord, py + v3.yCoord, pz + v3.zCoord, f21.D, f41.D)
			tessellator.addVertexWithUV(px + v4.xCoord, py + v4.yCoord, pz + v4.zCoord, f21.D, f5.D)
			tessellator.draw()
		}
	}
}